pkg/controllers/metrics/state/node.go

```diff
@@ -0,0 +1,250 @@
+		allocatable := n.Node.Status.Capacity
+		if len(n.Node.Status.Allocatable) > 0 {
+			allocatable = n.Node.Status.Allocatable
+		}
+
+		// Populate  metrics
+		var err error
+		for gaugeVec, resourceList := range map[*prometheus.GaugeVec]v1.ResourceList{
+			overheadGaugeVec:       ns.getSystemOverhead(n.Node),
+			podRequestsGaugeVec:    podRequests,
+			podLimitsGaugeVec:      podLimits,
+			daemonRequestsGaugeVec: n.DaemonSetRequested,
+			daemonLimitsGaugeVec:   n.DaemonSetLimits,
+			allocatableGaugeVec:    allocatable,
+		} {
+			err = multierr.Append(err, ns.set(resourceList, n.Node, gaugeVec))
+		}
+
+		if err != nil {
+			logging.FromContext(ctx).Errorf("Failed to generate gauges for %s: %s", n.Node.Name, err)
+		}
+
+		return true
+	})
+
+	ns.cleanup(ctx, nodes)
+}
+
+func (ns *nodeScraper) reset() {
+	for _, gauge := range []*prometheus.GaugeVec{
+		allocatableGaugeVec,
+		podRequestsGaugeVec,
+		podLimitsGaugeVec,
+		daemonRequestsGaugeVec,
+		daemonLimitsGaugeVec,
+		overheadGaugeVec,
+	} {
+		gauge.Reset()
+	}
+}
+
+func (ns *nodeScraper) cleanup(ctx context.Context, existingNodes map[string]struct{}) {
+	nodesToRemove := []string{}
+	for nodeName := range ns.labelMap {
+		if _, ok := existingNodes[nodeName]; !ok {
+			nodesToRemove = append(nodesToRemove, nodeName)
+		}
+	}
+
+	// Remove all gauges associated with removed node
+	for _, node := range nodesToRemove {
+		gaugeMap := ns.labelMap[node]
+		for gaugeVec, labelSet := range gaugeMap {
+			for _, labels := range labelSet {
+				gaugeVec.Delete(labels)
+			}
+		}
+		delete(ns.labelMap, node)
+	}
+
+	if len(nodesToRemove) > 0 {
+		logging.FromContext(ctx).Debugf("Removed the following node gauges: %s", strings.Join(nodesToRemove, ", "))
+	}
+}
+
+// set sets the value for the node gauge
+func (ns *nodeScraper) set(resourceList v1.ResourceList, node *v1.Node, gaugeVec *prometheus.GaugeVec) error {
+	for resourceName, quantity := range resourceList {
+		// Reformat resource type to be consistent with Prometheus naming conventions (snake_case)
+		resourceTypeName := strings.ReplaceAll(strings.ToLower(string(resourceName)), "-", "_")
+
+		labels := ns.getNodeLabels(node, resourceTypeName)
+		ns.labelMap[node.Name][gaugeVec] = append(ns.labelMap[node.Name][gaugeVec], labels)
+
+		gauge, err := gaugeVec.GetMetricWith(labels)
+		if err != nil {
+			return fmt.Errorf("generating new gauge: %w", err)
+		}
+		if resourceName == v1.ResourceCPU {
+			gauge.Set(float64(quantity.MilliValue()) / float64(1000))
+		} else {
+			gauge.Set(float64(quantity.Value()))
+		}
+	}
+	return nil
+}
+
+func (ns *nodeScraper) getSystemOverhead(node *v1.Node) v1.ResourceList {
+	systemOverheads := v1.ResourceList{}
+	if len(node.Status.Allocatable) > 0 {
+		// calculating system daemons overhead
+		for resourceName, quantity := range node.Status.Allocatable {
+			overhead := node.Status.Capacity[resourceName]
+			overhead.Sub(quantity)
```

Comment:
Is there any chance this is mutating the value? IIUC, this takes a pointer

Reply:
If you're referring to `overhead.Sub(quantity)`, `quantity` is passed by value.